pkg/documentlog/impl.go

```diff
@@ -73,112 +76,148 @@ func (dl *documentLog) Configure(publicAddr string) {
 func (dl *documentLog) Subscribe(documentType string) DocumentQueue {
 	queue := documentQueue{
 		documentType: documentType,
-		c:            make(chan *model.Document, 100), // TODO: Does this number make sense?
+		c:            make(chan model.Document, 100), // TODO: Does this number make sense?
 	}
 	dl.subscriptionsMutex.WriteLock(func() {
 		dl.subscriptions = append(dl.subscriptions, queue)
 	})
 	return &queue
 }
 
-func (dl documentLog) GetDocument(hash model.Hash) model.Document {
+func (dl documentLog) GetDocument(hash model.Hash) *model.Document {
+	var entry *entry
+	dl.documentsMutex.ReadLock(func() {
+		entry = dl.documentHashIndex[hash.String()]
+	})
+	if entry == nil {
+		return nil
+	}
+	return &entry.Document
+}
+
+func (dl documentLog) GetDocumentContents(hash model.Hash) (io.ReadCloser, error) {
 	var entry *entry
 	dl.documentsMutex.ReadLock(func() {
 		entry = dl.documentHashIndex[hash.String()]
 	})
-	if entry == nil || entry.doc == nil {
-		return model.Document{}
+	if entry == nil {
+		return nil, ErrUnknownDocument
+	}
+	// TODO: Nil/len check on Contents?
+	if entry.contents == nil {
+		return nil, ErrMissingDocumentContents
 	}
-	return *entry.doc
+	return NoopCloser{Reader: bytes.NewReader(*entry.contents)}, nil
 }
 
 func (dl *documentLog) HasDocument(hash model.Hash) bool {
+	// TODO: Implement this without a lock
+	var result bool
+	dl.documentsMutex.ReadLock(func() {
+		result = dl.documentHashIndex[hash.String()] != nil
+	})
+	return result
+}
+
+func (dl *documentLog) HasContentsForDocument(hash model.Hash) bool {
+	// TODO: Implement this without a lock
 	var result bool
 	dl.documentsMutex.ReadLock(func() {
-		result = dl.documentHashIndex[hash.String()].doc != nil
+		result = dl.documentHashIndex[hash.String()] != nil && dl.documentHashIndex[hash.String()].contents != nil
 	})
 	return result
 }
 
-func (dl *documentLog) AddDocumentHash(hash model.Hash, timestamp time.Time) {
+func (dl *documentLog) AddDocumentWithContents(document model.Document, contents io.Reader) error {
```

Comment:
If you only expose a func that can handle contents and then calculate the hash/timestamp yourself then less things can go wrong?

Reply:
(see comment above regarding whether requiring the caller to supply the hash.)